docs/src/main/asciidoc/optaplanner.adoc

```diff
@@ -446,8 +446,7 @@ public class TimeTableConstraintProvider implements ConstraintProvider {
 
         // Select a lesson ...
         return constraintFactory.from(Lesson.class)
-                // ... and pair it with another lesson ...
-                .join(Lesson.class,
+                .fromUniquePair(Lesson.class,
```

Comment:
`.from(Lesson.class).fromUniquePair(...)` will give a compile error

Reply:
Hi @ge0ffrey, thank you for reviewing. I updated the class. Following the current guide on site, it is breaking on tests.

Reply:
@rhuan080 Intresting, this does sound like an issue.
Which test exactly? The `TimeTableResourceTest` later down the guide? Or is the a ConstraintVerifier using test that I am missing?

Reply:
Hi @ge0ffrey,

I have checked this comment said by you -->fromUniquePair() is sugar for from(A).join(A, ...)) but for beginners especially those that know SQL, it's better to use from+join.

Looking at the source code, I think it has a small difference. The .fromUniquePair is filtering the planningId. It means just the planningId < PLANNING_ID_OF_FROM will match to it. I think it will generate a return with just ordered planningId. Thus I think the set of data returned from fromUniquePair() will be less than the set of data generated by from+join. Look at this code snippet:

return this.from(fromClass).join(fromClass, joiner, Joiners.lessThan(planningIdGetter));

Let me know if my thought is correct. If so, it explains the issue I have seen in the tests.

Reply:
@ge0ffrey updated. Thank you for helping me to understand the problem and solve it. :-)